modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,660 @@
+        buf.putLong(uuid.localId());
+    }
+
+    /** Reads {@link IgniteUuid} from buffer. */
+    public static IgniteUuid readIgniteUuid(ByteBuffer buf) {
+        UUID globalId = new UUID(buf.getLong(), buf.getLong());
+
+        return new IgniteUuid(globalId, buf.getLong());
+    }
+
+    /** Worker to write to performance statistics file. */
+    private class FileWriter extends GridWorker {
+        /** Maximum cached string count. */
+        private static final short MAX_CACHED_STRING_COUNT = Short.MAX_VALUE;
+
+        /** Performance statistics file I/O. */
+        private final FileIO fileIo;
+
+        /** File write buffer. */
+        private final SegmentedRingByteBuffer ringByteBuffer;
+
+        /** Minimal batch size to flush in bytes. */
+        private final int flushBatchSize;
+
+        /** Size of ready for flushing bytes. */
+        private final AtomicInteger readyForFlushSize = new AtomicInteger();
+
+        /** Stop file writer future. */
+        GridFutureAdapter<Void> stopFut = new GridFutureAdapter<>();
+
+        /** Cached strings by id. */
+        private final ConcurrentHashMap<String, Short> stringIds = new ConcurrentHashMap<>();
+
+        /** String id generator. */
+        private final AtomicInteger idsGen = new AtomicInteger();
+
+        /** {@code True} if the small buffer warning message logged. */
+        private final AtomicBoolean smallBufLogged = new AtomicBoolean();
+
+        /** {@code True} if worker stopped due to maximum file size reached. */
+        private final AtomicBoolean stopByMaxSize = new AtomicBoolean();
+
+        /**
+         * @param ctx Kernal context.
+         * @param fileIo Performance statistics file I/O.
+         * @param maxFileSize Maximum file size in bytes.
+         * @param bufferSize Off heap buffer size in bytes.
+         * @param flushBatchSize Minimal batch size to flush in bytes.
+         * @param log Logger.
+         */
+        FileWriter(GridKernalContext ctx, FileIO fileIo, long maxFileSize, int bufferSize, int flushBatchSize,
+            IgniteLogger log) {
+            super(ctx.igniteInstanceName(), "performance-statistics-writer%" + ctx.igniteInstanceName(), log);
+
+            this.fileIo = fileIo;
+            this.flushBatchSize = flushBatchSize;
+
+            ringByteBuffer = new SegmentedRingByteBuffer(bufferSize, maxFileSize, BufferMode.DIRECT);
+
+            ringByteBuffer.init(0);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void body() throws InterruptedException, IgniteInterruptedCheckedException {
+            while (!isCancelled() && !Thread.interrupted()) {
+                blockingSectionBegin();
+
+                try {
+                    synchronized (this) {
+                        while (readyForFlushSize.get() < flushBatchSize && !isCancelled())
+                            wait();
+                    }
+
+                    flushBuffer();
```

Comment:
Why `flushBuffer` is in blocking section?

Reply:
I have moved buffer flushing out of the blocking section.